src/main/java/com/slack/kaldb/server/ArmeriaService.java

```diff
@@ -0,0 +1,107 @@
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ArmeriaService extends AbstractIdleService {
+  private final Logger LOG = LoggerFactory.getLogger(ArmeriaService.class);
+
+  private final PrometheusMeterRegistry prometheusMeterRegistry;
+  private final KaldbServiceGrpc.KaldbServiceImplBase searcher;
+  private final String serviceName;
+  private final Server server;
+  private final int serverPort;
+
+  public ArmeriaService(
+      int serverPort,
+      PrometheusMeterRegistry prometheusMeterRegistry,
+      KaldbServiceGrpc.KaldbServiceImplBase searcher) {
+    this(serverPort, prometheusMeterRegistry, searcher, null);
+  }
+
+  public ArmeriaService(
+      int serverPort,
+      PrometheusMeterRegistry prometheusMeterRegistry,
+      KaldbServiceGrpc.KaldbServiceImplBase searcher,
+      String serviceName) {
+    this.serverPort = serverPort;
+    this.prometheusMeterRegistry = prometheusMeterRegistry;
+    this.searcher = searcher;
+    this.serviceName = serviceName;
+    this.server = getServer();
+  }
+
+  public Server getServer() {
+    ServerBuilder sb = Server.builder();
+    GrpcServiceBuilder searchBuilder =
+        GrpcService.builder().addService(searcher).enableUnframedRequests(true);
+    sb.service(searchBuilder.build());
+
+    sb.annotatedService(new ElasticsearchApiService(searcher));
+
+    addManagementEndpoints(sb, serverPort);
+
+    return sb.build();
+  }
+
+  private void addManagementEndpoints(ServerBuilder sb, int serverPort) {
+    sb.decorator(
+        MetricCollectingService.newDecorator(GrpcMeterIdPrefixFunction.of("grpc.service")));
+    sb.decorator(getLoggingServiceBuilder().newDecorator());
+    sb.http(serverPort);
+    sb.service("/health", HealthCheckService.builder().build());
+    sb.service("/metrics", (ctx, req) -> HttpResponse.of(prometheusMeterRegistry.scrape()));
+    sb.serviceUnder("/docs", new DocService());
+    sb.serviceUnder("/internal/management/", ManagementService.of());
+  }
+
+  private LoggingServiceBuilder getLoggingServiceBuilder() {
+    return LoggingService.builder()
+        // Not logging any successful response, say prom scraping /metrics every 30 seconds at INFO
+        .successfulResponseLogLevel(LogLevel.DEBUG)
+        .failureResponseLogLevel(LogLevel.ERROR)
+        // Remove all headers to be sure we aren't leaking any auth/cookie info
+        .requestHeadersSanitizer((ctx, headers) -> DefaultHttpHeaders.EMPTY_HEADERS);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    CompletableFuture<Void> serverFuture = server.start();
+    serverFuture.get(15, TimeUnit.SECONDS);
+    LOG.info("Started on port: {}", serverPort);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Shutting down");
+    server.closeAsync().get(15, TimeUnit.SECONDS);
```

Comment:
I think `stop` also returns a CompletableFuture

Looking at the armeria code for `stop` vs `closeAsync` I think `stop` does more? 

I definitely see more usages of stop - https://github.com/line/armeria/search?q=server.stop

Not a 100% sure what the subtle difference if at all b/w the two are

Reply:
This is a good question, and I had to do a bit more digging to find the differences between the two - https://github.com/line/armeria/issues/3598. 

`.stop()` is expected to be deprecated in favor of `.close()`/`.closeAsync()`. The current difference between the two seems to be that close cannot be restarted, where stopped can be.